### PR TITLE
[WIP] Ner pipeline grouped_entities fixes

### PR DESCRIPTION
**Pull request by cceyda**
There are many issues with ner pipeline using grouped_entities=True
https://github.com/huggingface/transformers/issues/5077
https://github.com/huggingface/transformers/issues/4816
https://github.com/huggingface/transformers/issues/5730
https://github.com/huggingface/transformers/issues/5609
https://github.com/huggingface/transformers/issues/6514
https://github.com/huggingface/transformers/issues/5541

- [x]  [Bug Fix] add an option `ignore_subwords` to ignore subsequent ##wordpieces in predictions. Because some models train on only the first token of a word and not on the subsequent wordpieces (BERT NER default). So it makes sense doing the same thing at inference time.

    - The simplest fix is to just group the subwords with the first wordpiece. 
      - [TODO] how to handle ignored scores? just set them to 0 and calculate zero invariant mean ?
      - [TODO] handle different wordpiece_prefix ## ? possible approaches:
get it from tokenizer? but currently most tokenizers dont have a wordpiece_prefix property?
have an _is_subword(token)

- [x] [Feature add] added option to `skip_special_tokens`. Cause It was harder to remove them after grouping.

- [x] [Additional Changes] remove B/I prefix on returned grouped_entities 

Edit: Ignored subwords' scores are also ignored by setting them to nan and using nanmean
Edit: B entities of different type are separated (as per BIO tag definition)
Edit: skip_special_tokens is now the default behavior
Edit: ignore_subwords is now the default behavior
Edit: more flexibility for custom non-standard tokenizers through tokenizer.is_subword_fn, tokenizer.convert_tokens_to_string
Edit: [fix UNK token related bugs by mapping UNK tokens to the correct original string] Use fast tokenizer or pass offset_mapping

# Usage
`pipeline('ner', model=model, tokenizer=tokenizer, ignore_labels=[], grouped_entities=True, ignore_subwords=True)`

### CODE REVIEW

**Review by enzoampil — CHANGES_REQUESTED** [review]
Thanks for this @cceyda , this generally looks good to me with just some comments around:

1. I think we can hard code `skip_special_tokens=True`, similar to the approach in `TextGenerationPipeline`.
2. I'm wondering about why we shouldn't group B type entities? Hoping you can elaborate your rationale here
3. Please make sure to add the test cases to `NerPipelineTests` in [test_pipelines](https://github.com/huggingface/transformers/blob/master/tests/test_pipelines.py). These should be similar test cases to the ones you referenced in this PR (that used to fail, but now pass with these changes).
4. Please update the tests to account for these changes. The current failures seem to be due to removal of the prefixes in "entity type".

Thanks!

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py:1132]: `B-` is supposed to signify the start of a new entity by the definition of [IOB2 format](https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging)), which is commonly used for example by CONLL ner task.
(Ah this was supposed to be in a separate [PR](https://github.com/huggingface/transformers/pull/5966)...but I guess I miss-commited)

Comment (enzoampil) [reply src/transformers/pipelines.py:1132]: I understand. I think this is reasonable, I'm just thinking of potential cases where an entity is labelled `B-` at the middle of a token. For example, in this recent [issue](https://github.com/huggingface/transformers/issues/4816) where `B-` is used for two consecutive word pieces. For these cases, I think the word pieces should still be grouped.

Perhaps we can account for this added complexity in a separate PR from this one then? I'll suggest to remove this line first.

```
[
{'word': 'Cons', 'score': 0.9994944930076599, 'entity': 'B-PER', 'index': 1},
{'word': '##uelo', 'score': 0.802545428276062, 'entity': 'B-PER', 'index': 2}
]
```

File: src/transformers/pipelines.py
Comment (enzoampil) [inline src/transformers/pipelines.py:1129]:
```suggestion
```

File: src/transformers/pipelines.py
Comment (enzoampil) [inline src/transformers/pipelines.py:1132]:
```suggestion
```

**Comment by HHoofs** [comment]
I'm wondering should the B & I part maybe separated from the entity type part? In the sense that you average the entities (disregarding the B/I part) and vice-versa. I now have the feeling that only the first subtoken decides whether the complete word is a B or an I.

**Comment by cceyda** [comment]
I want to complete this but ran into another issue while working it:

All [UNK] tokens get mapped to [UNK] in the output, instead of the actual input token (because the code is getting from ids->tokens), Also [UNK]s gets lost when using skip_special_tokens (https://github.com/huggingface/transformers/issues/6863)
While this is a simple token alignment issue and can be solved by using `offset_mappings`. `offset_mappings` is only available with fast tokenizers, I'm wondering what would be a more general approach to solving this?

**Comment by Monique497** [comment]
Dear @cceyda,

In the last couple of days I started to work with transformers and especially NER-classification. I ran into issues that has been previously addressed in other issues you just mentioned at the beginning. Especially that subtokens that were classified with 'O'  were not properly merged with the full token. 

For example (Dutch):
sentence = "Als we Volkswagens OR-voorzitter **Bernd Osterloh** moeten geloven, dan moet dat binnen drie jaar het geval zijn."

Gives me as group entities: 
[{'entity_group': 'B-per', 'score': 0.9999980926513672, 'word': 'Bern'}, 
{'entity_group': 'I-per', 'score': 0.9999990463256836, 'word': 'Ost'}]

I expect:
[{'entity_group': 'B-per', 'score': 0.9999980926513672, 'word': 'Bernd'}, 
{'entity_group': 'I-per', 'score': 0.9999990463256836, 'word': 'Osterloh'}]

However, the considered subtokens are classified as 'O':

{'word': '[CLS]', 'score': 0.9999999403953552, 'entity': 'O', 'index': 0}
{'word': 'Als', 'score': 0.9999999403953552, 'entity': 'O', 'index': 1}
{'word': 'we', 'score': 0.9999999403953552, 'entity': 'O', 'index': 2}
{'word': 'Volkswagen', 'score': 0.9999955296516418, 'entity': 'B-misc', 'index': 3}
{'word': '##s', 'score': 0.9999999403953552, 'entity': 'O', 'index': 4}
{'word': 'O', 'score': 0.9981945157051086, 'entity': 'I-misc', 'index': 5}
{'word': '##R', 'score': 0.9999998807907104, 'entity': 'O', 'index': 6}
{'word': '-', 'score': 0.9999999403953552, 'entity': 'O', 'index': 7}
{'word': 'voorzitter', 'score': 0.9999998807907104, 'entity': 'O', 'index': 8}
{'word': 'Bern', 'score': 0.9999980926513672, 'entity': 'B-per', 'index': 9}
**{'word': '##d', 'score': 0.9999998807907104, 'entity': 'O', 'index': 10}**
{'word': 'Ost', 'score': 0.9999990463256836, 'entity': 'I-per', 'index': 11}
**{'word': '##er', 'score': 0.9999998807907104, 'entity': 'O', 'index': 12}
{'word': '##lo', 'score': 0.9999997615814209, 'entity': 'O', 'index': 13}
{'word': '##h', 'score': 0.9999998807907104, 'entity': 'O', 'index': 14}**
{'word': 'moeten', 'score': 0.9999999403953552, 'entity': 'O', 'index': 15}
{'word': 'geloven', 'score': 0.9999998807907104, 'entity': 'O', 'index': 16}
{'word': ',', 'score': 0.9999999403953552, 'entity': 'O', 'index': 17}
{'word': 'dan', 'score': 0.9999999403953552, 'entity': 'O', 'index': 18}
{'word': 'moet', 'score': 0.9999999403953552, 'entity': 'O', 'index': 19}
{'word': 'dat', 'score': 0.9999999403953552, 'entity': 'O', 'index': 20}
{'word': 'binnen', 'score': 0.9999999403953552, 'entity': 'O', 'index': 21}
{'word': 'drie', 'score': 0.9999999403953552, 'entity': 'O', 'index': 22}
{'word': 'jaar', 'score': 0.9999999403953552, 'entity': 'O', 'index': 23}
{'word': 'het', 'score': 0.9999999403953552, 'entity': 'O', 'index': 24}
{'word': 'geval', 'score': 0.9999999403953552, 'entity': 'O', 'index': 25}
{'word': 'zijn', 'score': 0.9999999403953552, 'entity': 'O', 'index': 26}
{'word': '.', 'score': 0.9999999403953552, 'entity': 'O', 'index': 27}
{'word': '[SEP]', 'score': 0.9999999403953552, 'entity': 'O', 'index': 28}

I believe your pull request addresses these issues properly. 
However, I saw the merge did not complete since it failed on some tasks.

I was wondering if there is still the intention to solve these issues.

Disclaimer: I am a total newbie to git (just set up an account), so please be mild, haha.
Any help is much appreciated!

Thank you in advance,

Monique

**Comment by enzoampil** [comment]
@cceyda I actually want this PR to move forward. Are you okay collaborating on your fork (can add me as collaborator)? I can help out with some of the issues failing so we can get this merged :smile:

**Comment by cceyda** [comment]
@enzoampil I have added you as a collaborator.
Also pushed some additional changes addressing the [UNK] token mapping problem I mentioned before. 
Still there are some things I'm not very satisfied with:

1. subword prefix was fixed to '##' before. with the latest change I added a check to see if the tokenizer has an `is_subword_fn `defined (still dont like handling it this way). I know some tokenizers have `subword_prefix` but most don't and this was the most flexible solution for now. 
2. `offset_mappings`  is needed to resolve [UNK] tokens, but is only available with fast tokenizers. Fast tokenizers don't have `convert_ids_to_tokens` so had to implement a hacky solution for those aswell.
3.  `skip_special_tokens` also dropped [UNK] tokens so I had to change things and rely on `special_tokens_mask`.

It is not optimal but it worked for my use cases. 
Haven't had a chance to look at the failing tests yet :/

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py]: Now I'm thinking maybe this should be optional?

**Comment by cceyda** [comment]
I have changed the `ignore_subwords` default to True which covers cases like
```
[
{'word': 'Cons', 'score': 0.9994944930076599, 'entity': 'B-PER', 'index': 1},
{'word': '##uelo', 'score': 0.802545428276062, 'entity': 'B-PER', 'index': 2}
]
```
And honestly I don't know why subwords shouldn't be ignored for most cases. (Unless there is need for some custom logic that determines a words tag; ie by averaging the wordpieces etc etc. In which case grouped_entities shouldn't be used 🤔  )
IMO Mid-word inconsistencies made by the model while `ignore_subwords = False` shouldn't effect pipelines output logic.

[todo]
- torch tests are passing for now but probably should add more cases? (I can't see why the tf tests are failing though, don't have dev env for that)
- should add the new parameters to the doc strings.

**Comment by Monique497** [comment]
This seems good, now, doesn't it? :) Thank you for picking this up, again. Me and my team really appreciate that.

**Comment by Monique497** [comment]
Dear @cceyda,

Last two days I worked on your branch to see how it performs on my own input texts. 
However, I  came accross the following issue I would like to point out to you:

When I use the following line of code (as you suggest under 'Usage' above):

pipeline('ner', model=model, tokenizer=tokenizer, ignore_labels=[], grouped_entities=True, skip_special_tokens=True, ignore_subwords=True)


I get the error:

TypeError: __init__() got an unexpected keyword argument 'skip_special_tokens'.

When looking in the file transformer.pipelines and looking specifically for the tokenclassificationpipeline, it seems that it is not yet implemented. Or am I missing something?

Best, 

Monique

**Comment by cceyda** [comment]
@Monique497 sorry for the delay
A couple of things have changed since I first wrote that example:
 
- special tokens ([CLS][PAD][SEP]) are always skipped (per comments above) so you don't need that kwarg.

``` py
from transformers import (
    AutoModelForTokenClassification,
    AutoTokenizer,
    pipeline,
)

model = AutoModelForTokenClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True) # note the fast tokenizer use
# ignore_subwords = True by default
nlp = pipeline("ner",model=model,tokenizer=tokenizer, grouped_entities=True)
inputs="test sentence"
output=nlp(inputs)
```

- Another important thing is you have to **use a fast tokenizer** OR pass `offset_mapping` as a parameter because the [UNK] token resolution depends on this. (maybe I should rename this to offset_mappings)

```py
# you can pass it like this
nlp(inputs,offset_mapping=mappings_you_calculate)

```
- If you are using a custom tokenizer that treats subwords differently (ie not starting with '##'), you can pass a function implementing your custom logic through `tokenizer.is_subword_fn`  and `tokenizer.convert_tokens_to_string` 
I don't know if this is the best way to handle non standard tokenizations, but I use some custom non-standard tokenizers for Korean and this solution gave me enough flexibility.

something like this:

```py
def sub_fn(token):
    if token.starts_with("%%"): return True
tokenizer.is_subword_fn=sub_fn

def convert_tokens_to_string(self, tokens):
    out_string = " ".join(tokens).replace(" %%", "").strip()
    return out_string
tokenizer.convert_tokens_to_string=convert_tokens_to_string
```

@enzoampil what are your thoughts on this?

File: tests/test_pipelines.py
Comment (cceyda) [inline tests/test_pipelines.py:788]: conditioned so that grouped_entities=False tests won't fail because of grouped_entities=True

File: tests/test_pipelines.py
Comment (cceyda) [inline tests/test_pipelines.py:789]: added case for ignore_subwords=True and False

**Review by cceyda — COMMENTED** [review]
What else can I do for this pr for merge? It has been a while

**Comment by enzoampil** [comment]
@cceyda Sorry for taking a while, lemme do another review!

File: src/transformers/pipelines.py
Comment (enzoampil) [reply src/transformers/pipelines.py]: @cceyda Actually, I reviewed again, and agree we should keep this here.

File: tests/test_pipelines.py
Comment (enzoampil) [reply tests/test_pipelines.py:788]: LGTM

**Review by enzoampil — APPROVED** [review]
Awesome work @cceyda ! I think this looks good to go :smile:

**Comment by enzoampil** [comment]
@LysandreJik @julien-c This looks good to me. Please advise if this is good to merge or if you think there's still anything missing before merging :grin:

File: src/transformers/pipelines.py
Comment (LysandreJik) [inline src/transformers/pipelines.py:1396]: This seems like it would break if ran from a tensorflow pipeline

File: src/transformers/pipelines.py
Comment (LysandreJik) [inline src/transformers/pipelines.py:1474]: This is very specific to BERT

File: src/transformers/pipelines.py
Comment (LysandreJik) [inline src/transformers/pipelines.py:1483]: This too is very specific to BERT

**Review by LysandreJik — CHANGES_REQUESTED** [review]
Thanks for iterating on this! There seems to be an issue with the TF tests, as it doesn't seem to be testing the pipeline in itself (see comment).

Also, this approach uses very BERT-specific splitting/joining, which we can't have as we should aim to support all 30 models.

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py:1474]: I provide alternatives for those BERT specific functions; the user can override the function by providing their own tokenizer.is_subword_fn and tokenizer.convert_tokens_to_string. while it is not a clean solution it is flexible. I would like it if there was a property like tokenizer.subword_prefix for all tokenizers, then I could use that. (I remember I saw something similar somewhere maybe in the fast tokenizers?)

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py:1483]: same here. can be overridden by tokenizer.convert_tokens_to_string. I recently saw the [BatchEncoding.token_to_word](https://huggingface.co/transformers/main_classes/tokenizer.html#transformers.BatchEncoding.token_to_word) maybe I can use that, will take a look

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py:1396]: oh I see that, will fix

Comment (LysandreJik) [reply src/transformers/pipelines.py:1396]: Alright, thanks for making sure it works!

File: src/transformers/pipelines.py
Comment (LysandreJik) [reply src/transformers/pipelines.py:1474]: I'm pretty sure for slow tokenizers one can use `tokenizer.convert_tokens_to_string`, while for fast tokenizer you can use `token_to_word` to obtain the indices and construct words.

I don't think expecting the user to provide a method to join tokens is the best we can do here. Pipelines are meant as a way for users to use the models without needing to understand what happens under the hood, having them specify a tokenizer-specific method to join tokens goes against that goal.

File: src/transformers/pipelines.py
Comment (LysandreJik) [reply src/transformers/pipelines.py:1483]: Can we use the `token_to_word` if it's available, `convert_tokens_to_string` otherwise?

File: tests/test_pipelines.py
Comment (jgibson) [inline tests/test_pipelines.py:743]: A test for hyphenated names (ex. Juantia Gomez-Cortez) would be useful, especially given that the fast and slow tokenizers have different codepaths for reconstructing the original text. I had to implement grouping of named entities myself recently and was tripped up by that corner case.

File: src/transformers/pipelines.py
Comment (cceyda) [inline src/transformers/pipelines.py:1458]: changed `is_subword` detection logic: by comparing length of token(##token) with the original text span mapping (Assuming subwordpieces get prefixed by something). 
Incase the user wants some other logic they can first get ungrouped entities add `is_subword:bool` field to entities and call `pipeline.group_entities` themselves.

File: src/transformers/pipelines.py
Comment (cceyda) [inline src/transformers/pipelines.py:1501]: fixed as suggested! I agree it is much cleaner this way.

File: src/transformers/pipelines.py
Comment (cceyda) [inline src/transformers/pipelines.py:1302]: added this to check `offset_mapping` if provided. (does a simple batch_size check)

File: src/transformers/pipelines.py
Comment (cceyda) [reply src/transformers/pipelines.py]: I agree about the pipelines, hadn't liked my previous hacky solution myself either. I have changed tokenizer.convert_tokens_to_string bit as suggested https://github.com/huggingface/transformers/pull/5970#discussion_r511790482. and also the subword detection logic.https://github.com/huggingface/transformers/pull/5970#discussion_r511789502

File: tests/test_pipelines_ner.py
Comment (cceyda) [inline tests/test_pipelines_ner.py:162]: Don't know why the tests are failing, I get normal results when running them outside of the test suite 😕 ?

**Comment by LysandreJik** [comment]
Thanks for iterating! I'll check this today.

**Review by LysandreJik — APPROVED** [review]
Played around with it, works well, and the implementation seems robust. Looks good, LGTM! Thanks for iterating.

@stefan-it, do you want to take a quick look?

**Comment by LysandreJik** [comment]
Merging this as soon as it's green, thank you for iterating on the PR! Sorry this took so long to merge.

**Comment by enzoampil** [comment]
Thanks @LysandreJik and congrats @cceyda !! :smile:

**Comment by Botfacke** [comment]
fix_bart_gpu

**Comment by LysandreJik** [comment]
FYI this broke the behavior of the NER pipeline:

```py
from transformers import pipeline

nlp = pipeline("ner")

nlp("My name is Alex and I live in New York")
```

crashes with the following error:
```
    raise Exception("To decode [UNK] tokens use a fast tokenizer or provide offset_mapping parameter")
Exception: To decode [UNK] tokens use a fast tokenizer or provide offset_mapping parameter
```

Trying to see if this can be quickly patched, otherwise we'll revert the PR while we patch this.

**Comment by cceyda** [comment]
oops! although returning unk tokens with slow tokenizers are not the best, I agree not forcing a fast tokenizer with a default of ignore_subword=True looks better for keeping the compatibility. I saw a bit late the _args_parser line was mis-merged during this pr merge and I see it is fixed/improved on the patch. I wasn't sure on how to test for the offset_mapping argument with the new test structure (which looks to be good at the patch). Sorry for the trouble 😅 @LysandreJik

**Comment by LysandreJik** [comment]
No worries, thanks for taking a look at the patch!

**Pull request merged**
